### PR TITLE
Animation Refactor

### PR DESCRIPTION
**Pull request by nickdichev**
I thought it would be fun to try to implement the game "simon says" on the Xebow, but I found some things missing from the animation code! Primarily, the ability to play "one shot" animations, eg: flash these keys in this order.

I did a bit of a refactor here to add support for "one shot" and looping pre-defined animations.

Additionally, for "dynamic" animations (the ones which were already there), I added the ability to track the last few frames of animation data. I think this will let us implement more complex effects like an LED "wave" that reflects off the edge of the matrix and other key presses.

### CODE REVIEW

File: lib/xebow/rgb_matrix.ex [hunk 1]
```diff
@@ -50,8 +50,8 @@ defmodule Xebow.RGBMatrix do
     GenServer.start_link(__MODULE__, [], name: __MODULE__)
   end
 
-  def flash(color) do
-    GenServer.cast(__MODULE__, {:flash, color})
+  def flash(color, sleep_ms \\ 250) do
```
[the remark this answers is not in the document]
Comment (nickdichev) [reply lib/xebow/rgb_matrix.ex:36]: Will do, I like the idea of turning this into an animation!

File: lib/xebow/animation.ex [hunk 2]
```diff
@@ -40,13 +66,16 @@ defmodule Xebow.Animation do
...
         %Animation{
           type: animation_type,
           tick: opts[:tick] || 0,
           speed: opts[:speed] || 100,
           delay_ms: opts[:delay_ms] || 17,
-          pixels: pixels,
-          pixel_colors: opts[:pixel_colors] || init_pixel_colors(pixels)
+          next_frame: nil,
+          frames: [init_frame],
+          frame_buffer_length: opts[:frame_buffer_length] || 3
```
[the remark this answers is not in the document]
Comment (nickdichev) [reply lib/xebow/animation.ex:78]: My idea was that some animations might need/want more previous frame data for their effects. I tend to agree that this is mostly a bolted on after thought though and will remove the buffer until the wave animation is thought through a little more.

File: lib/xebow/animation.ex [hunk 3]
```diff
@@ -57,6 +86,10 @@ defmodule Xebow.Animation do
         Enum.map(pixels, fn _pixel -> Chameleon.HSV.new(0, 0, 0) end)
       end
 
+      # We don't want the length of the frame buffer to continuously grow, however, we do
+      # want to store some of the frames so animations can use previous frames as reference.
+      defp buffer_frames(frames, buffer_len), do: Enum.take(frames, buffer_len)
```
[the remark this answers is not in the document]
Comment (nickdichev) [reply lib/xebow/animation.ex:91]: Will remove

File: lib/xebow/animation_frame.ex [hunk 4]
```diff
@@ -0,0 +1,29 @@
+defmodule Xebow.AnimationFrame do
```
[the remark this answers is not in the document]
Comment (nickdichev) [reply lib/xebow/animation_frame.ex:1]: Will do.

Interesting idea for a rendering engine on top of the `Animation` abstraction. I had a thought that maybe a protocol could be used, something like:

```elixir
defprotocol Renderable do
  @spec render(animation :: %Xebow.Animation{}) :: spi_color_map()
  def render(animation)
end
```

```elixir
defimpl Renderable, for: Any do
  def render(animation) do
      animation.frame.pixel_map
      |> Enum.sort()
      |> Enum.map(fn {_cord, color} -> color end)

    data =
      Enum.reduce(colors, @sof, fn color, acc ->
        rgb = Chameleon.convert(color, Chameleon.Color.RGB)
        acc <> <<227, rgb.b, rgb.g, rgb.r>>
      end) <> @eof
  end
end
```

```elixir
defmodule Xebow.Animation do
  @derive Renderable
  defstruct [...]
end
```

I dunno though, this is pretty coupled to the fact that we're using SPI as an interface to the device?

File: lib/xebow/animation_frame.ex [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  @moduledoc false
```
[the remark this answers is not in the document]
Comment (nickdichev) [reply lib/xebow/animation_frame.ex:2]: Will add :)

File: lib/xebow/animation.ex [hunk 6]
```diff
@@ -21,15 +35,27 @@ defmodule Xebow.Animation do
...
       end
 
-      # Increment the animation state to the next tick.
-      @spec do_tick(animation :: Animation.t()) :: Animation.t()
-      defp do_tick(animation) do
-        %Animation{animation | tick: animation.tick + 1}
+      @impl Animation
+      @spec next_state(animation :: Animation.t()) :: Animation.t()
+      # Predefined animations
+      def next_state(%{frame_buffer_length: nil} = animation) do
```
[the remark this answers is not in the document]
Comment (nickdichev) [reply lib/xebow/animation.ex:46]: Sounds good, will do.

File: lib/xebow/rgb_matrix.ex [hunk 7]
```diff
@@ -62,9 +45,25 @@ defmodule Xebow.RGBMatrix do
     GenServer.cast(__MODULE__, :previous_animation)
   end
 
+  @spec play_one_shot(animation :: Animation.t()) :: :ok
+  def play_one_shot(animation) do
```
[the remark this answers is not in the document]
Comment (nickdichev) [reply lib/xebow/rgb_matrix.ex:49]: I like the idea of adding a `:loop` field to the struct! Will do

Not really following on the idea of "eating ticks", will need to follow up on this maybe at the next planning meeting :)

File: lib/xebow/animation_frame.ex (hunk 4 above)
[the remark this answers is not in the document]
Comment (amclain) [reply lib/xebow/animation_frame.ex:1]: >this is pretty coupled to the fact that we're using SPI

Yeah, I'd lean towards decoupling. I'm not sure we've documented this yet, but one cool idea @doughsay brought up is to be able to network multiple Keybows together and render an animation across all of them (i.e. add more pixels to the RGB matrix). If animations are decoupled from the RGB matrix hopefully it will make implementing this feature easier. In other words, conversion to SPI sounds like a good responsibility of the `RGBMatrix` module.

Things like this sound like a good topic for an architecture discussion so that we can ensure the major pieces of code being refactored support the use cases envisioned for Xebow.

File: lib/xebow/rgb_matrix.ex (hunk 7 above)
[the remark this answers is not in the document]
Comment (amclain) [reply lib/xebow/rgb_matrix.ex:49]: I probably didn't explain that well. You've seen how animations typically increment the tick when they move on to the next state:

https://github.com/ElixirSeattle/xebow/blob/e1532fe0c23daaf6b81405043cc953e479ba4db2/lib/xebow/animation/cycle_all.ex#L24-L25

Well for animations that are dynamic, rather than using the `tick` to determine their state, they could return their next state without needing to increment the tick. In this sense they're "eating the tick" -- discarding it.

This could definitely be a good topic for one of the planning meetings, and it might touch on when a buffer of previously rendered frames would be helpful, which you brought up. I'd love to discuss it with you at a planning meeting. Hope to see you there!

**Comment by amclain** [comment]
I don't necessarily want the following thought to sidetrack this PR, but I was going to bring it up at tonight's planning meeting. Since the meeting was cancelled I feel like I should at least mention it here to make you aware of it.

There may be a UX painpoint to rearchitecting Xebow animations around frames depending on how we do it. I was reminded of this when making the animated gifs for #26. If an animation developer needs to define every frame of an animation for animations that are "bitmap-like" (or "non-procedural"), it could get incredibly tedious, especially for high frame rate animations. The good news is that this is a solved problem and animation editing software typically uses [keyframes](https://en.wikipedia.org/wiki/Key_frame) and [tweens](https://en.wikipedia.org/wiki/Inbetweening). The question is if we would want to go that route with Xebow or try to do something else. I haven't had time to see how [QMK](https://qmk.fm/) handles this yet either; not sure if @doughsay knows.

File: lib/xebow/rgb_matrix.ex (hunk 1 above)
[the remark this answers is not in the document]
Comment (nickdichev) [reply lib/xebow/rgb_matrix.ex]: So take a look at what I have in the most recent revision. Really, a flash is a single frame of animation played for some duration. So I implemented this with the new `Xebow.Animation.Static` animation type.

File: lib/xebow/animation.ex (hunk 2 above)
[the remark this answers is not in the document]
Comment (nickdichev) [reply lib/xebow/animation.ex]: This was removed.

File: lib/xebow/animation.ex (hunk 3 above)
[the remark this answers is not in the document]
Comment (nickdichev) [reply lib/xebow/animation.ex]: This was removed.

File: lib/xebow/animation_frame.ex (hunk 4 above)
[the remark this answers is not in the document]
Comment (nickdichev) [reply lib/xebow/animation_frame.ex]: Renamed `Xebow.AnimationFrame` -> `Xebow.Frame`.

File: lib/xebow/rgb_matrix.ex (hunk 7 above)
[the remark this answers is not in the document]
Comment (nickdichev) [reply lib/xebow/rgb_matrix.ex]: This worked out very nicely. I re-implemented the one-shot types to use the `:loop` field on the `%Xebow.Animation{}` and to use the same `play_animation()` API from the `Xebow.RGBMatrix`. 

One side effect of this is that the one-shots are now executed as a cast. I see a use case where the caller _may_ want to be blocked while the animation is playing (simon says for example, want to ignore input while animation is playing). Do you think adding a `call` flow is worthwhile, or should the caller be expected to sleep?

File: lib/xebow/animation.ex (hunk 6 above)
[the remark this answers is not in the document]
Comment (nickdichev) [reply lib/xebow/animation.ex]: Removed the entire concept of state from this guy. The new contract is a lot simpler with only frames!

File: lib/xebow/animation_frame.ex (hunk 5 above)
[the remark this answers is not in the document]
Comment (nickdichev) [reply lib/xebow/animation_frame.ex]: Added documentation here and updated the other docs throughout. Let me know if something looks out of place since I might have missed something in this second revision.

File: lib/xebow/rgb_matrix.ex (hunk 7 above)
[the remark this answers is not in the document]
Comment (nickdichev) [reply lib/xebow/rgb_matrix.ex]: Also definitely looking forward to chatting at the next planning meeting!

File: lib/xebow/animation/static.ex [hunk 8]
```diff
@@ -0,0 +1,114 @@
+defmodule Xebow.Animation.Static do
```
Comment (nickdichev) [inline lib/xebow/animation/static.ex:1]: This is a pretty bad name for what it represents... (a static animation? lol). Was thinking maybe "predefined" fits better? I dunno in other places (mostly docs) I have been calling the other animation types "dynamic" animations so I went with "Static" for these guys.

File: lib/xebow/animation.ex [hunk 9]
```diff
@@ -4,11 +4,12 @@ defmodule Xebow.Animation do
 
   There are currently two distinct ways to define an animation.
 
-  You may define an animation with a predefined `:frames` field. Each frame will advance every `:delay_ms` milliseconds. These animations should use the `Xebow.Animation.Loop` `:type`. See the moduledoc of `Xebow.Animation.Loop` for an example. These animations should have a nil `:frame_buffer_length`.
+  You may define an animation with a predefined `:frames` field. Each frame will advance every `:delay_ms` milliseconds.
+  These animations should use the `Xebow.Animation.{OneShot,Loop}` `:type`. See the moduledocs of those modules for
```
Comment (amclain) [inline lib/xebow/animation.ex:8]: Sorry, small nitpick: It looks like `{OneShot,Loop}` is missing a space after the comma.

File: lib/xebow/animation.ex [hunk 10]
```diff
@@ -22,11 +22,12 @@ defmodule Xebow.Animation do
           type: type,
           tick: non_neg_integer,
           speed: non_neg_integer,
+          loop: non_neg_integer | -1,
```
Comment (amclain) [inline lib/xebow/animation.ex]: Your typing here reminded me of something I should have thought of sooner: What do you think about `loop: non_neg_integer | :infinitely`?  I realized the atom may be more self-documenting than `-1`, so I'm curious how you feel about it.

File: lib/xebow/animation.ex [hunk 11]
```diff
@@ -13,67 +13,30 @@ defmodule Xebow.Animation do
   """
 
   alias __MODULE__
-  alias Xebow.{AnimationFrame, RGBMatrix}
+  alias Xebow.{AnimationFrame}
```
Comment (amclain) [inline lib/xebow/animation.ex]: I don't think the curly braces are needed here anymore if there's only one module to import, unless I'm overlooking something.

File: lib/xebow/animation/static.ex (hunk 8 above)
Comment (amclain) [reply lib/xebow/animation/static.ex:1]: :thinking: I think `Static` works. Maybe @doughsay has an opinion on this one.

File: lib/xebow/rgb_matrix.ex (hunk 7 above)
[the remark this answers is not in the document]
Comment (amclain) [reply lib/xebow/rgb_matrix.ex]: >Do you think adding a `call` flow is worthwhile, or should the caller be expected to sleep?

My instinct says you're on the right track with _not_ blocking the caller during an animation. What do you think about a callback when the animation completes? That may lead to the need for a `wait_for_animation()` convenience function in the caller, which could circle back around to a `cast` being simpler and utilizing built-in GenServer mechanics. You may have gone down this path already; what are your thoughts?

[the remark this answers is not in the document]
Comment (nickdichev) [reply lib/xebow/rgb_matrix.ex]: I haven't played around with this yet. I think by default playing an animation should not block the caller but there should be a way to block while an animation is playing. I do think that using the GenServer builtins would be simpler and avoid "reinventing the wheel". 

I think this is a good candidate topic for the next planning meeting :)

File: lib/xebow/animation.ex (hunk 11 above)
Comment (nickdichev) [reply lib/xebow/animation.ex]: hmm I don't know how you saw this -- I definitely fixed this in one of my commits from yesterday. This is just `alias Xebow.Frame` on the latest commit.

File: lib/xebow/animation.ex (hunk 9 above)
Comment (nickdichev) [reply lib/xebow/animation.ex]: Good catch, this should actually be `Xebow.Animation.Static`, fixed.

File: lib/xebow/animation.ex (hunk 10 above)
Comment (nickdichev) [reply lib/xebow/animation.ex]: Yes very good. I was debating this when implementing but thought -1 makes sense as well since its definitely an invalid `:loop` value. The atom reads much nicer in function head pattern matching though.

Changed the `:loop` field to use `:infinite` to represent dynamic animations.

File: lib/xebow/animation.ex (hunk 11 above)
Comment (amclain) [reply lib/xebow/animation.ex]: :thinking: Sorry, ignore me. I believe I was browsing through the commits, late at night, saw `alias Xebow.{AnimationFrame}` in one of them, and commented in the commit rather than at the end of the PR. My bad.

File: lib/xebow/animation/static.ex (hunk 8 above)
Comment (amclain) [inline lib/xebow/animation/static.ex:1]: I noticed a regression in the red and green static animations. When keys 12 + 1 or 12 + 3 are double tapped quickly to trigger the static animations, they get latched in the on state. I haven't traced the code on this one yet so I'm starting this convo on the module so it's easier to discuss than in the PR's main comment section.

Comment (nickdichev) [reply lib/xebow/animation/static.ex:1]: Indeed. Good catch.

I added the option to make `play_animation/1` synchronous and made the keys server use that option!

Comment (amclain) [reply lib/xebow/animation/static.ex:1]: Awesome, thanks!

File: lib/xebow/keys.ex [hunk 12]
```diff
@@ -156,12 +157,14 @@ defmodule Xebow.Keys do
 
   # Custom Key Functions
 
-  def flash_red do
-    Xebow.RGBMatrix.flash("red")
-  end
+  def flash(color) do
```
Comment (amclain) [inline lib/xebow/keys.ex:160]: (no action)

Nice refactor! I like how you pulled `flash` out of the `RGBMatrix` and moved it into the "application". At some point we should probably separate the application layer (the key presses and demo logic) from the library/hardware. For example, the demo's animations should probably be separated from the key handling code. But I think that can be a problem for another PR.

**Comment by doughsay** [comment]
Sorry I've been so MIA lately... I'll take a look at this tonight I swear!

File: lib/xebow/animation/static.ex [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+  @moduledoc """
+  Pre-defined animations that run as a one-shot or on a loop.
+
+  The `Xebow.Animation.Static` animation type is used to define animations with a pre-defined set of frames. These
+  animations can be played continiously or on a loop. This behavior is controlled by the `:loop` field on the animation.
+  A value of :infinite means to play the animation continuously, while a value greater than 0 means to loop the animation
```
Comment (doughsay) [inline lib/xebow/animation/static.ex:7]:
```suggestion
  A value of `:infinite` means to play the animation continuously, while a value greater than 0 means to loop the animation
```

Comment (nickdichev) [reply lib/xebow/animation/static.ex]: I fixed it!

[the remark this answers is not in the document]
Comment (nickdichev) [reply lib/xebow/animation/static.ex]: I fixed it!

File: lib/xebow/animation.ex [hunk 14]
```diff
@@ -78,19 +54,58 @@ defmodule Xebow.Animation do
...
+          {:type, animation_type}
+          | {:frames, list}
+          | {:tick, non_neg_integer}
+          | {:speed, non_neg_integer}
+          | {:delay_ms, non_neg_integer}
+          | {:loop, non_neg_integer | :infinite}
+
+  @type animation_opts :: list(animation_opt)
+
+  @spec new(opts :: animation_opts) :: Animation.t()
```
[the remark this answers is not in the document]
Comment (nickdichev) [reply lib/xebow/animation.ex]: Indeed! I made this change.
